File: server/engine/recommender.js
// Simple Rule-Based Recommendation Engine

function generateRecommendation(farm, iot, drone, kbCrops, kbDiseases) {
    let recommendation = {
        crop_suggestion: "Keep monitoring",
        water_advice: "Check water levels",
        disease_detected: "None",
        medicine_suggestion: "None",
        medicine_secondary: "None",
        dosage: "N/A",
        preventive_measures: "Monitor regularly",
        timeline: "Monitor regularly",
        // New fields for UI
        confidence: 0,
        severity: "Low",
        image_reference: null,
        annotated_image_reference: null // If available
    };

    // 1. Crop Suggestion
    if (farm.current_crop === 'None' || farm.current_crop === '') {
        const soilRec = kbCrops.find(c => c.soil_type === farm.soil_type);
        if (soilRec) {
            recommendation.crop_suggestion = `Recommended: ${soilRec.recommended_crop}`;
        }
    } else {
        recommendation.crop_suggestion = `Current Crop: ${farm.current_crop}`;
    }

    // 2. Water Advice
    if (iot) {
        if (iot.water_level < 5.0) {
            recommendation.water_advice = "water_low";
        } else if (iot.water_level > 10.0) {
            recommendation.water_advice = "water_high";
        } else {
            recommendation.water_advice = "water_optimal";
        }
    }

    // 3. Disease & Medicine & Image Data
    if (drone) {
        recommendation.disease_detected = drone.disease_type;
        recommendation.confidence = drone.confidence || 0;
        recommendation.severity = drone.severity || "Unknown";
        // Ensure paths are web-accessible if they are relative
        recommendation.image_reference = drone.image_reference;
        recommendation.annotated_image_reference = drone.annotated_image_reference;

        // Normalize for lookup
        const normalize = (str) => str.toLowerCase().replace(/_/g, ' ').trim();
        const target = normalize(drone.disease_type);

        const diseaseRule = kbDiseases.find(d =>
            normalize(d.disease_name) === target ||
            normalize(d.disease_name).includes(target) ||
            target.includes(normalize(d.disease_name))
        );

        if (diseaseRule) {
            recommendation.medicine_suggestion = diseaseRule.medicine;
            recommendation.medicine_secondary = diseaseRule.medicine_secondary;
            recommendation.dosage = diseaseRule.dosage;
            recommendation.preventive_measures = diseaseRule.preventive_measures;
            recommendation.timeline = diseaseRule.timeline;
        } else if (drone.disease_type !== 'Healthy' && drone.disease_type !== 'Normal') {
            recommendation.medicine_suggestion = "Consult local expert";
            recommendation.dosage = "N/A";
            recommendation.preventive_measures = "Quarantine affected area";
            recommendation.timeline = "Immediate action required";
        }
    }

    return recommendation;
}

module.exports = { generateRecommendation };

File: server/engine/recommender.py
def generate_recommendation(farm: dict, iot: dict, drone: dict, kb_crops: list, kb_diseases: list) -> dict:
    recommendation = {
        "crop_suggestion": "Keep monitoring",
        "water_advice": "Check water levels",
        "disease_detected": "None",
        "medicine_suggestion": "None",
        "medicine_secondary": "None",
        "dosage": "N/A",
        "preventive_measures": "Monitor regularly",
        "timeline": "Monitor regularly",
        "confidence": 0,
        "severity": "Low",
        "image_reference": None,
        "annotated_image_reference": None
    }

    current_crop = farm.get("current_crop", "")
    if current_crop == "None" or current_crop == "":
        soil_rec = next((c for c in kb_crops if c.get("soil_type") == farm.get("soil_type")), None)
        if soil_rec:
            recommendation["crop_suggestion"] = f"Recommended: {soil_rec.get('recommended_crop')}"
    else:
        recommendation["crop_suggestion"] = f"Current Crop: {current_crop}"

    if iot:
        water_level = iot.get("water_level", 0.0)
        if water_level < 5.0:
            recommendation["water_advice"] = "water_low"
        elif water_level > 10.0:
            recommendation["water_advice"] = "water_high"
        else:
            recommendation["water_advice"] = "water_optimal"

    if drone:
        disease_type = drone.get("disease_type", "")
        recommendation["disease_detected"] = disease_type
        recommendation["confidence"] = drone.get("confidence", 0)
        recommendation["severity"] = drone.get("severity", "Unknown")
        recommendation["image_reference"] = drone.get("image_reference")
        recommendation["annotated_image_reference"] = drone.get("annotated_image_reference")

        def normalize(s: str) -> str:
            return s.lower().replace("_", " ").strip() if s else ""

        target = normalize(disease_type)
        
        disease_rule = None
        for d in kb_diseases:
            d_name = normalize(d.get("disease_name", ""))
            if d_name == target or target in d_name or d_name in target:
                disease_rule = d
                break

        if disease_rule:
            recommendation["medicine_suggestion"] = disease_rule.get("medicine")
            recommendation["medicine_secondary"] = disease_rule.get("medicine_secondary")
            recommendation["dosage"] = disease_rule.get("dosage")
            recommendation["preventive_measures"] = disease_rule.get("preventive_measures")
            recommendation["timeline"] = disease_rule.get("timeline")
        elif disease_type != 'Healthy' and disease_type != 'Normal':
            recommendation["medicine_suggestion"] = "Consult local expert"
            recommendation["dosage"] = "N/A"
            recommendation["preventive_measures"] = "Quarantine affected area"
            recommendation["timeline"] = "Immediate action required"

    return recommendation
